include/pybind11/cast.h

```diff
@@ -1002,11 +1002,47 @@ template <> class type_caster<std::nullptr_t> : public void_caster<std::nullptr_
 
 template <> class type_caster<bool> {
 public:
-    bool load(handle src, bool) {
+    bool load(handle src, bool convert) {
         if (!src) return false;
         else if (src.ptr() == Py_True) { value = true; return true; }
         else if (src.ptr() == Py_False) { value = false; return true; }
-        else return false;
+        else if (convert) {
+            // This is quite similar to PyObject_IsTrue(), but it doesn't default
+            // to "True" for arbitrary objects.
+            Py_ssize_t res = -1;
+            if (src.is_none()) {
+                res = 0;
+            }
+            #if !defined(PYPY_VERSION)
+            #if PY_MAJOR_VERSION >= 3
+            else if (src.ptr()->ob_type->tp_as_number && src.ptr()->ob_type->tp_as_number->nb_bool) {
+                res = (*src.ptr()->ob_type->tp_as_number->nb_bool)(src.ptr());
+            }
+            #else
+            else if (src.ptr()->ob_type->tp_as_number && src.ptr()->ob_type->tp_as_number->nb_nonzero) {
+                res = (*src.ptr()->ob_type->tp_as_number->nb_nonzero)(src.ptr());
+            }
```

Comment:
You could clean this up quite a bit (9 lines -> 3) by adding a `#define PYBIND11_NB_BOOL nb_bool` into common.h and collapsing the two `else if`s into 1 without needing to `#ifdef` on the Python version.